File: src/page/info/edit/SCGroupEditTable.java
package page.info.edit;

import java.awt.Component;
import java.awt.event.KeyEvent;
import java.util.EventObject;

import javax.swing.text.JTextComponent;

import io.Reader;
import page.MainLocale;
import page.support.AbJTable;
import page.support.EnemyTCR;
import util.Data;
import util.stage.SCDef;
import util.stage.SCGroup;
import util.unit.AbEnemy;
import util.unit.EnemyStore;

class SCGroupEditTable extends AbJTable {

	private static final long serialVersionUID = 1L;

	private static String[] title;

	static {
		redefine();
	}

	protected static void redefine() {
		title = MainLocale.getLoc(1, "t1", "t8");
	}

	protected final SCDef scd;

	protected SCGroupEditTable(SCDef sc) {
		scd = sc;
		setDefaultRenderer(Integer.class, new EnemyTCR());
	}

	@Override
	public boolean editCellAt(int row, int column, EventObject e) {
		boolean result = super.editCellAt(row, column, e);
		Component editor = getEditorComponent();
		if (editor == null || !(editor instanceof JTextComponent))
			return result;
		if (e instanceof KeyEvent)
			((JTextComponent) editor).selectAll();
		return result;
	}

	@Override
	public Class<?> getColumnClass(int c) {
		return lnk[c] == 0 ? Integer.class : String.class;
	}

	@Override
	public int getColumnCount() {
		return title.length;
	}

	@Override
	public String getColumnName(int c) {
		return title[lnk[c]];
	}

	@Override
	public synchronized int getRowCount() {
		if (scd == null)
			return 0;
		return scd.smap.size();
	}

	@Override
	public synchronized Object getValueAt(int r, int c) {
		if (scd == null || r < 0 || c < 0 || r >= scd.smap.size() || c > lnk.length)
			return null;
		return get(r, lnk[c]);
	}

	@Override
	public boolean isCellEditable(int r, int c) {
		return true;
	}

	@Override
	public synchronized void setValueAt(Object arg0, int r, int c) {
		if (scd == null)
			return;
		c = lnk[c];
		if (c > 0) {
			int[] is = Reader.parseIntsN((String) arg0);
			if (is.length == 0)
				return;
			if (is.length == 1)
				set(r, c, is[0], -1);
			else
				set(r, c, is[0], is[1]);
		} else {
			int i = arg0 instanceof Integer ? (Integer) arg0 : Reader.parseIntN((String) arg0);
			set(r, c, i, 0);
		}
	}

	protected synchronized void addLine(AbEnemy enemy) {
		if (scd == null)
			return;
		int ind = getSelectedRow();
		int eid = 0;
		if (enemy == null)
			while (scd.smap.containsKey(eid++))
				;
		else
			eid = enemy.getID();
		if (scd.smap.containsKey(eid))
			return;
		scd.smap.put(eid, 0);
		ind++;
		addRowSelectionInterval(ind, ind);
	}

	protected synchronized void remLine() {
		if (scd == null)
			return;
		int ind = getSelectedRow();
		if (ind == -1)
			return;
		scd.smap.remove(scd.getSMap()[ind][0]);
		if (ind >= scd.smap.size())
			ind--;
		addRowSelectionInterval(ind, ind);
	}

	private Object get(int r, int c) {
		int[][] info = scd.getSMap();
		if (r >= info.length)
			return null;
		int[] data = info[r];
		if (c == 0)
			return EnemyStore.getAbEnemy(data[0], true);
		else if (c == 1) {
			int g = data[1];
			SCGroup scg = scd.sub.get(g);
			return scg == null ? g != 0 ? Data.trio(g) + " - invalid" : "" : scg.toString();
		}
		return null;
	}

	private void set(int r, int c, int v, int para) {
		int[][] info = scd.getSMap();
		if (r >= info.length)
			return;
		int[] data = info[r];
		if (v < 0)
			v = 0;
		if (c == 0)
			scd.smap.put(v, scd.smap.remove(data[0]));
		else if (c == 1)
			scd.smap.put(data[0], v);
	}

}

File: src/page/info/edit/SCGroupEditTable.kt
package page.info.edit

import common.CommonStatic
import common.pack.PackData
import common.util.Data
import common.util.stage.SCDef
import common.util.stage.SCGroup
import common.util.unit.AbEnemy
import page.MainLocale
import page.support.AbJTable
import page.support.EnemyTCR
import java.awt.Component
import java.awt.event.KeyEvent
import java.util.*
import javax.swing.text.JTextComponent

internal class SCGroupEditTable(sc: SCDef?) : AbJTable() {
    companion object {
        private const val serialVersionUID = 1L
        private var title: Array<String>
        fun redefine() {
            title = MainLocale.Companion.getLoc(1, "t1", "t8")
        }

        init {
            redefine()
        }
    }

    protected val scd: SCDef?
    override fun editCellAt(r: Int, c: Int, e: EventObject): Boolean {
        val result: Boolean = super.editCellAt(r, c, e)
        val editor: Component = getEditorComponent()
        if (editor == null || editor !is JTextComponent) return result
        val jtf: JTextComponent = editor as JTextComponent
        if (e is KeyEvent) jtf.selectAll()
        return result
    }

    override fun getColumnClass(c: Int): Class<*> {
        return if (lnk.get(c) == 0) Int::class.java else String::class.java
    }

    override fun getColumnCount(): Int {
        return title.size
    }

    override fun getColumnName(c: Int): String {
        return title[lnk.get(c)]
    }

    @Synchronized
    override fun getRowCount(): Int {
        return if (scd == null) 0 else scd.smap.size
    }

    @Synchronized
    override fun getValueAt(r: Int, c: Int): Any {
        return if (scd == null || r < 0 || c < 0 || r >= scd.smap.size || c > lnk.size) null else get(r, lnk.get(c))!!
    }

    override fun isCellEditable(r: Int, c: Int): Boolean {
        return lnk.get(c) != 0
    }

    @Synchronized
    override fun setValueAt(arg0: Any, r: Int, c: Int) {
        var c = c
        if (scd == null) return
        c = lnk.get(c)
        if (c == 1) {
            val `is`: IntArray = CommonStatic.parseIntsN(arg0 as String)
            if (`is`.size == 1) set(r, `is`[0])
        }
    }

    @Synchronized
    fun addLine(enemy: AbEnemy?) {
        if (scd == null) return
        var ind: Int = getSelectedRow()
        if (enemy == null) return
        val eid: PackData.Identifier<AbEnemy> = enemy.getID()
        if (scd.smap.containsKey(eid)) return
        scd.smap.put(eid, 0)
        ind++
        if (ind < 0) clearSelection() else setRowSelectionInterval(ind, ind)
    }

    @Synchronized
    fun remLine() {
        if (scd == null) return
        var ind: Int = getSelectedRow()
        if (ind == -1) return
        scd.smap.remove(scd.getSMap().get(ind).key)
        if (ind >= scd.smap.size) ind--
        if (ind < 0) clearSelection() else setRowSelectionInterval(ind, ind)
    }

    private operator fun get(r: Int, c: Int): Any? {
        val info: Array<Map.Entry<PackData.Identifier<AbEnemy>, Int>> = scd.getSMap()
        if (r >= info.size) return null
        if (c == 0) return info[r].key.get() else if (c == 1) {
            val g = info[r].value
            val scg: SCGroup = scd.sub.get(g)
            return if (scg == null) if (g != 0) Data.Companion.trio(g) + " - invalid" else "" else scg.toString()
        }
        return null
    }

    private operator fun set(r: Int, v: Int) {
        var v = v
        val info: Array<Map.Entry<PackData.Identifier<AbEnemy>, Int>> = scd.getSMap()
        if (r >= info.size) return
        val data: Map.Entry<PackData.Identifier<AbEnemy>, Int> = info[r]
        if (v < 0) v = 0
        scd.smap.put(data.key, v)
    }

    init {
        scd = sc
        setDefaultRenderer(Int::class.java, EnemyTCR())
    }
}
